src/sql/src/plan/statement/ddl.rs

```diff
@@ -1151,9 +1136,53 @@ pub fn plan_create_source(
         seen: _,
     } = CreateSourceOptionExtracted::try_from(with_options.clone())?;
 
-    let (key_desc, value_desc) = encoding.desc()?;
+    let encoding = match format {
+        Some(format) => Some(get_encoding(scx, format, &envelope)?),
+        None => None,
+    };
 
-    let mut key_envelope = get_key_envelope(include_metadata, &encoding)?;
+    let (key_desc, value_desc) = match &encoding {
+        Some(encoding) => {
+            // If we are applying an encoding we need to ensure that the incoming value_desc is a
+            // single column of type bytes.
+            match external_connection.value_desc().typ().columns() {
+                [typ] => match typ.scalar_type {
+                    ScalarType::Bytes => {}
+                    _ => sql_bail!(
```

Comment:
I think this error message is likely to make sense to MZ contributors but wonder about its legibility to new users.

Feel free to disregard or season to your taste, but threw together a diff to improve the error message:

```diff
diff --git a/src/sql/src/plan/error.rs b/src/sql/src/plan/error.rs
index ee419b1d5d..432b96c2c6 100644
--- a/src/sql/src/plan/error.rs
+++ b/src/sql/src/plan/error.rs
@@ -252,6 +252,10 @@ pub enum PlanError {
         /// created
         hypothetical_replica_count: usize,
     },
+    IncompatibleSourceEncoding {
+        source_name: &'static str,
+        format: mz_sql_parser::ast::CreateSourceFormat<crate::names::Aug>,
+    },
     // TODO(benesch): eventually all errors should be structured.
     Unstructured(String),
 }
@@ -429,6 +433,9 @@ impl PlanError {
             Self::SubsourceNameConflict { .. } | Self::SubsourceAlreadyReferredTo { .. } => {
                 Some("Specify target table names using FOR TABLES (foo AS bar), or limit the upstream tables using FOR SCHEMAS (foo)".into())
             },
+            Self::IncompatibleSourceEncoding {.. } => {
+                Some("For compatibility between connections and formats, see https://materialize.com/docs/sql/create-source/".into())
+            }
             _ => None,
         }
     }
@@ -685,6 +692,9 @@ impl fmt::Display for PlanError {
             Self::CreateReplicaFailStorageObjects {..} => {
                 write!(f, "cannot create more than one replica of a cluster containing sources or sinks")
             },
+            Self::IncompatibleSourceEncoding { source_name, format } => {
+                write!(f, "the specified {} connection is incompatible with{}", source_name, format.to_ast_string())
+            }
         }
     }
 }
diff --git a/src/sql/src/plan/statement/ddl.rs b/src/sql/src/plan/statement/ddl.rs
index f60b269671..3bf824a903 100644
--- a/src/sql/src/plan/statement/ddl.rs
+++ b/src/sql/src/plan/statement/ddl.rs
@@ -1146,15 +1146,21 @@ pub fn plan_create_source(
             // If we are applying an encoding we need to ensure that the incoming value_desc is a
             // single column of type bytes.
             match external_connection.value_desc().typ().columns() {
-                [typ] => match typ.scalar_type {
-                    ScalarType::Bytes => {}
-                    _ => sql_bail!(
-                        "The schema produced by the source is incompatible with format decoding"
-                    ),
-                },
-                _ => sql_bail!(
-                    "The schema produced by the source is incompatible with format decoding"
-                ),
+                [ColumnType {
+                    scalar_type: ScalarType::Bytes,
+                    ..
+                }] => {}
+                _ => {
+                    let format = match format {
+                        Some(format) => format.clone(),
+                        None => sql_bail!("[internal error] encoding is some but format is none"),
+                    };
+
+                    return Err(PlanError::IncompatibleSourceEncoding {
+                        source_name: external_connection.name(),
+                        format,
+                    });
+                }
             }
 
             let (key_desc, value_desc) = encoding.desc()?;
```

Reply:
Thanks for this diff, I incorporated it!